snowplow-common-internal/kinesis/src/main/scala/com.snowplowanalytics.snowplow/sources/KinesisSource.scala

```diff
@@ -0,0 +1,252 @@
+                // We must not release the semaphore yet, because we are not ready for fs2-aws to checkpoint the end of the shard.
+                // We can safely ignore the exception and move on.
+                Logger[F].warn(
+                  s"Checkpointing failed on a record which was not the last in the shard. Meanwhile, KCL has already enqueued the final record in the shard to the fs2 queue. Ignoring error and instead continue processing towards the shard end"
+                )
+            }
+        }
+    def nack(c: List[KinesisMetadata[F]]): F[Unit] = Applicative[F].unit
+  }
+
+  private def kinesisStream[F[_]: Async](config: KinesisSourceConfig): Stream[F, LowLevelEvents[List[KinesisMetadata[F]]]] = {
+    val region = Region.of(config.region.getOrElse(getRuntimeRegion))
+    val resources =
+      for {
+        consumerSettings <- Resource.pure[F, KinesisConsumerSettings](
+                              KinesisConsumerSettings(
+                                config.streamName,
+                                config.appName,
+                                region,
+                                bufferSize = PosInt.unsafeFrom(config.bufferSize)
+                              )
+                            )
+        kinesisClient <- mkKinesisClient[F](region, config.customEndpoint)
+        dynamoClient <- mkDynamoDbClient[F](region, config.dynamodbCustomEndpoint)
+        cloudWatchClient <- mkCloudWatchClient[F](region, config.cloudwatchCustomEndpoint)
+        kinesis <- Resource.pure[F, Kinesis[F]](
+                     Kinesis.create(scheduler(kinesisClient, dynamoClient, cloudWatchClient, config, _))
+                   )
+      } yield (consumerSettings, kinesis)
+
+    Stream
+      .resource(resources)
+      .flatMap { case (settings, kinesis) =>
+        kinesis.readFromKinesisStream(settings)
+      }
+      .chunks
+      .map { chunk =>
+        LowLevelEvents(chunk.toList.map(getPayload), chunk.toList.map(toMetadata))
+      }
+  }
+
+  def getPayload(record: CommittableRecord): Array[Byte] = {
+    val data = record.record.data
+    val buffer = ArrayBuffer[Byte]()
+    while (data.hasRemaining)
+      buffer.append(data.get)
+    buffer.toArray
+  }
+
+  def getRuntimeRegion: String =
+    (new DefaultAwsRegionProviderChain).getRegion.id()
+
+  private def scheduler[F[_]: Sync](
+    kinesisClient: KinesisAsyncClient,
+    dynamoDbClient: DynamoDbAsyncClient,
+    cloudWatchClient: CloudWatchAsyncClient,
+    kinesisConfig: KinesisSourceConfig,
+    recordProcessorFactory: ShardRecordProcessorFactory
+  ): F[Scheduler] =
+    Sync[F].delay(UUID.randomUUID()).map { uuid =>
+      val hostname = InetAddress.getLocalHost().getCanonicalHostName()
+
+      val configsBuilder =
+        new ConfigsBuilder(
+          kinesisConfig.streamName,
+          kinesisConfig.appName,
+          kinesisClient,
+          dynamoDbClient,
+          cloudWatchClient,
+          s"$hostname:$uuid",
```

Comment:
Do we need the hostname here?  I checked what [fs2-aws does by default](https://github.com/laserdisc-io/fs2-aws/blob/v6.0.3/fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/Kinesis.scala#L146) does by default and it simply uses a uuid.